### PR TITLE
More flexible gdp adapter

### LINKED ISSUE #94: Allow setting a download directory for individual GDP files
See https://github.com/Cloud-Drift/clouddrift/pull/88#discussion_r1089775668.

### PR DESCRIPTION
**Pull request by milancurcic**
Fixes #94 and #235.

### CODE REVIEW

**Comment by milancurcic** [comment]
CI on Ubuntu and macOS has been failing since yesterday due to segmentation fault in codecov (tests otherwise pass). Windows is OK. I haven't found anything about this from a cursory search. It could be transient or not.

@selipot unless you really want codecov to succeed, I think we can safely merge.

**Comment by philippemiron** [comment]
What happened to all the tests?

**Comment by milancurcic** [comment]
@philippemiron see https://github.com/Cloud-Drift/clouddrift/pull/237#issuecomment-1693354094. Codecov is crashing in CI. It started as yesterday as far as I know.

**Comment by philippemiron** [comment]
> @philippemiron see [#237 (comment)](https://github.com/Cloud-Drift/clouddrift/pull/237#issuecomment-1693354094). Codecov is crashing in CI. It started as yesterday as far as I know.

I see.. and it's passing for Windows. 😕

**Comment by milancurcic** [comment]
For once something works in Windows :)

**Comment by philippemiron** [comment]
I feel a Microsoft bias in here.. haha

**Review by selipot — COMMENTED** [review]
I have been able to test the new features such as `tmp_path` and the testing for existing files, all good
I noticed the following issues:
- using the FTP source 2.00 the ragged array is filled typically with one drifter less than the requested `n_random_id`. That behavior does not exist for experimental so I think it's ok.
- I cannot create a parquet file: it requires to use `to_awkward` from the ragged array instance and this fails because clouddrift.raggedarray.to_akward expects "count" instead of "rowsize"

**Comment by milancurcic** [comment]
I'd say just keep it in this PR for simplicity.

**Comment by milancurcic** [comment]
> I'd like to know if it resolves the issue of not being able to generate the parquet file via awkward

Hmm, works for me. What about it doesn't work for you? Is there an error message?

File: clouddrift/adapters/gdp1h.py [hunk 1]
```diff
@@ -60,17 +63,20 @@ def download(
         Randomly select n_random_id drifter IDs to download (Default: None)
     url : str
         URL from which to download the data (Default: GDP_DATA_URL). Alternatively, it can be GDP_DATA_URL_EXPERIMENTAL.
+    tmp_path : str, optional
+        Path to the directory where the individual NetCDF files are stored
+        (default /tmp/clouddrift/gdp)
```
[the remark this answers is not in the document]
Comment (milancurcic) [reply clouddrift/adapters/gdp1h.py]: I don't know but I just pushed a commit that clarifies that the default path varies on OS and this is on Linux.

**Comment by milancurcic** [comment]
> I cannot create a parquet file: it requires to use to_awkward from the ragged array instance and this fails because clouddrift.raggedarray.to_akward expects "count" instead of "rowsize"

@selipot A-ha, I see this now. OK, I think your commit worked, because I can write a parquet file via awkward.

**Comment by selipot** [comment]
> > I'd like to know if it resolves the issue of not being able to generate the parquet file via awkward
> 
> Hmm, works for me. What about it doesn't work for you? Is there an error message?

Wrote earlier: For some reason I can't get my environment to recognize my [latest commit](https://github.com/Cloud-Drift/clouddrift/pull/237/commits/445ea2bdc732a65870ce0d1a4f0e7f4e2f1272bd) and test it.

So I am not that skilled at VS code!

File: clouddrift/adapters/gdp1h.py [hunk 2]
```diff
@@ -65,7 +65,7 @@ def download(
         URL from which to download the data (Default: GDP_DATA_URL). Alternatively, it can be GDP_DATA_URL_EXPERIMENTAL.
     tmp_path : str, optional
         Path to the directory where the individual NetCDF files are stored
-        (default /tmp/clouddrift/gdp)
+        (default varies on OS; /tmp/clouddrift/gdp on Linux)
```
Comment (selipot) [inline clouddrift/adapters/gdp1h.py:68]: default varies depending on operating system

**Review by selipot — CHANGES_REQUESTED** [review]

File: clouddrift/adapters/gdp1h.py [hunk 3]
```diff
@@ -60,17 +63,20 @@ def download(
         Randomly select n_random_id drifter IDs to download (Default: None)
     url : str
         URL from which to download the data (Default: GDP_DATA_URL). Alternatively, it can be GDP_DATA_URL_EXPERIMENTAL.
+    tmp_path : str, optional
+        Path to the directory where the individual NetCDF files are stored
+        (default varies on OS; /tmp/clouddrift/gdp on Linux)
```
Comment (milancurcic) [inline clouddrift/adapters/gdp1h.py:68]:
```suggestion
        (default varies depending on operating system; /tmp/clouddrift/gdp on Linux)
```

File: clouddrift/adapters/gdp1h.py [hunk 4]
```diff
@@ -506,6 +513,9 @@ def to_raggedarray(
     url : str, optional
         URL from which to download the data (Default: GDP_DATA_URL).
         Alternatively, it can be GDP_DATA_URL_EXPERIMENTAL.
+    tmp_path : str, optional
+        Path to the directory where the individual NetCDF files are stored
+        (default /tmp/clouddrift/gdp)
```
Comment (milancurcic) [inline clouddrift/adapters/gdp1h.py:518]:
```suggestion
        (default varies depending on operating system; /tmp/clouddrift/gdp on Linux)
```

File: clouddrift/adapters/gdp6h.py [hunk 5]
```diff
@@ -47,17 +50,20 @@ def download(
         Randomly select n_random_id drifter IDs to download (Default: None)
     url : str
         URL from which to download the data (Default: GDP_DATA_URL). Alternatively, it can be GDP_DATA_URL_EXPERIMENTAL.
+    tmp_path : str, optional
+        Path to the directory where the individual NetCDF files are stored
+        (default varies on OS; /tmp/clouddrift/gdp6h on Linux)
```
Comment (milancurcic) [inline clouddrift/adapters/gdp6h.py:55]:
```suggestion
        (default varies depending on operating system; /tmp/clouddrift/gdp6h on Linux)
```

File: clouddrift/adapters/gdp6h.py [hunk 6]
```diff
@@ -434,6 +438,9 @@ def to_raggedarray(
         List of drifters to retrieve (Default: all)
     n_random_id : list[int], optional
         Randomly select n_random_id drifter NetCDF files
+    tmp_path : str, optional
+        Path to the directory where the individual NetCDF files are stored
+        (default /tmp/clouddrift/gdp6h)
```
Comment (milancurcic) [inline clouddrift/adapters/gdp6h.py:443]:
```suggestion
        (default varies depending on operating system; /tmp/clouddrift/gdp6h on Linux)
```

**Comment by selipot** [comment]
The tests fail because of `rowsize`

**Pull request merged**